### PR TITLE
Add dataset places365 with documentation and tests.

### LINKED ISSUE #132: [Dataset] Add Places365 Datasets for image classification
Please add the following Places Dataset as new _Image classification_ datasets: 

- [ ] [Places365](http://places2.csail.mit.edu/index.html)

Please ensure consistency with previous supported dataset with
- [ ] Reuse of the existing download and cache
- [ ] Equivalent test suite
- [ ] Documentation update

### PR DESCRIPTION
**Pull request by koshtiakanksha**
closes #132

### CODE REVIEW

**Comment by koshtiakanksha** [comment]
Hi @cregouby, the Places365 dataset implementation is now ready for review. Let me know if you'd like any further refinements or additions!

File: R/dataset-places365.R [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      self$samples <- lapply(files, function(p) {
+        parts <- fs::path_split(p)[[1]]
+        list(path = p, label = parts[length(parts) - 1])
+      })
+      classes <- sort(unique(vapply(self$samples, function(s) s$label, character(1))))
+      self$class_to_idx <- setNames(seq_along(classes), classes)
+    } else if (self$split == "val") {
+      lines <- suppressWarnings(readLines(self$val_ann))
+      parsed <- strsplit(lines, " ")
+      self$samples <- lapply(parsed, function(x) list(path = file.path(self$val_dir, x[[1]]), label = as.integer(x[[2]]) + 1))
```
Comment (cregouby) [inline R/dataset-places365.R:76]: **todo** Never use sample as name if you don't sample.
**todo** please avoid the raw readLines followed by unclear and unvectorized split function
**improvement** don't use suppressWarning. This is a code smell. You have to properly deal with errors.
**suggestion** rather use the efficient and straightforward `read.delim()` instead
**suggestion** data.frame format ensure you get the same aligned columns.
```suggestion
            ann <- read.delim(self$val_ann, header = FALSE, col.names = c("path", "label"), sep = " ")
            ann["label"] <- ann["label"] + 1L
            self$ann <- ann
```

File: R/dataset-places365.R
Comment (cregouby) [inline R/dataset-places365.R:1]: **todo** please use download_and_cache for any download.
**todo** please check files MD5 for any file downloaded
**todo** please use standardized user messages

**Review by cregouby — CHANGES_REQUESTED** [review]
**todo** a lot of improvements have been made lately on the way to download and manage datasets content. Please build on those improvements.
**todo** a lot of improvements have been made lately onstandardizing end-user messages. Please build on those improvements.

File: R/dataset-places365.R [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+#'
+#' @family classification_dataset
+#' @export
+
+places365_dataset <- torch::dataset(
+  name = "places365_dataset",
+  archive_size = "24 GB",
+
+  devkit_url = "http://data.csail.mit.edu/places/places365/filelist_places365-standard.tar",
+  devkit_md5 = "1a930ef26e7794c2c53d6c5f8c3c43d2",
```
Comment (cregouby) [inline R/dataset-places365.R:43]: **todo** should be changed into "35a0585fee1fa656440f3ab298f8479c"

File: R/dataset-places365.R [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+  .getitem = function(index) {
+    if (self$split == "test") {
+      path <- self$files[[index]]
+      if (!fs::file_exists(path))
+        cli_abort("Image file does not exist: {path}")
+      x <- magick::image_read(path)
+      if (!is.null(self$transform))
+        x <- self$transform(x)
+      list(x = x, y = NA_integer_)
```
Comment (cregouby) [inline R/dataset-places365.R:110]: **todo** y cannot be NA for a trainset. Please load annotation form `file.path(self$base_dir, "places365_train_standard.txt")`

File: R/dataset-places365.R [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    } else {
+      item <- self$items[[index]]
+      if (!fs::file_exists(item$path))
+        cli_abort("Image file does not exist: {item$path}")
+      x <- magick::image_read(item$path)
+      if (!is.null(self$transform))
+        x <- self$transform(x)
+      y <- if (self$split == "train") self$class_to_idx[[item$label]] else item$label
```
Comment (cregouby) [inline R/dataset-places365.R:118]: **todo** please revisit the logic. self$split cannot be "train" here as per L103 and L111

File: R/dataset-places365.R [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      length(self$files)
+    else
+      length(self$items)
+  },
+
+  .getitem = function(index) {
+    if (self$split == "test") {
+      path <- self$files[[index]]
+      if (!fs::file_exists(path))
+        cli_abort("Image file does not exist: {path}")
```
Comment (cregouby) [inline R/dataset-places365.R:106]: **todo** there will be nobody to look at those message as .getitem will be run inside a dataloader inside a torch model training loop. Please revise the logic so that it is filtered upfront (if needed)

**Review by cregouby — CHANGES_REQUESTED** [review]
**todo** there is still a poor coherency in the code desing compared to previous achievement. I guess a strong pass for a leaner code is needed.

File: R/dataset-places365.R [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+places365_dataset <- torch::dataset(
+  name = "places365_dataset",
+  archive_size = "24 GB",
+
+  devkit_url = "http://data.csail.mit.edu/places/places365/filelist_places365-standard.tar",
+  devkit_md5 = "1a930ef26e7794c2c53d6c5f8c3c43d2",
+
+  val_url = "http://data.csail.mit.edu/places/places365/val_256.tar",
+  val_md5 = "c50ef2655c708abf8bfa9e523101c4ec",
```
Comment (cregouby) [inline R/dataset-places365.R:46]: **todo** actual MD5 is "e27b17d8d44f4af9a78502beb927f808"

**Review by cregouby — CHANGES_REQUESTED** [review]
**todo** there is still a poor coherency in the code desing compared to previous achievement. I guess a strong pass for a leaner code is needed.
**todo** Please run the tests. They are currently failing. Implementing test skip is for CI-CD pipeline, you should run the tests once succesfully.

File: R/dataset-places365.R [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        y <- self$target_transform(y)
+      list(x = x, y = y)
+    }
+  },
+
+  download = function() {
+    if (self$check_exists())
+      return()
+
+    cli_inform("Downloading {.cls {class(self)[[1]]}} split: '{self$split}'")
```
Comment (cregouby) [inline R/dataset-places365.R:129]: **todo** please use the `archive_size` value and use the existing standard message as in any other dataset.

**Comment by koshtiakanksha** [comment]
Hi @tdhock this PR is now ready for review. I’ve incorporated all the suggested changes. Please let me know if you’d like any further modifications. Happy to make updates as needed.

File: man/places365_dataset.Rd [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+\arguments{
+\item{root}{Root directory for dataset storage. The dataset will be stored under \verb{root/<dataset-name>}. Defaults to \code{tempdir()}.}
+
+\item{split}{One of \code{"train"}, \code{"val"}, or \code{"test"}.}
+
+\item{transform}{Optional. A function that takes an image and returns a transformed version (e.g., normalization, cropping).}
+
+\item{target_transform}{Optional. A function that transforms the label.}
+
+\item{download}{Logical. If TRUE, downloads the dataset to \verb{root/}. If the dataset is already present, download is skipped.}
```
Comment (tdhock) [inline man/places365_dataset.Rd:24]: I see only 5 documented arguments here but 7 on the python page https://docs.pytorch.org/vision/main/generated/torchvision.datasets.Places365.html
```
root (str or pathlib.Path) – Root directory of the Places365 dataset.

        split (string, optional) – The dataset split. Can be one of train-standard (default), train-challenge, val, test.

        small ([bool](https://docs.python.org/3/library/functions.html#bool), optional) – If True, uses the small images, i.e. resized to 256 x 256 pixels, instead of the high resolution ones.

        download ([bool](https://docs.python.org/3/library/functions.html#bool), optional) – If True, downloads the dataset components and places them in root. Already downloaded archives are not downloaded again.

        transform (callable, optional) – A function/transform that takes in a PIL image and returns a transformed version. E.g, transforms.RandomCrop

        target_transform (callable, optional) – A function/transform that takes in the target and transforms it.

        loader – A function to load an image given its path.
```
the missing ones are small and loader. Can they be added please?

Comment (tdhock) [reply man/places365_dataset.Rd:26]: the arguments in R code are the same names as in python code, but not in the same order. Can you please change the order to be consistent?

File: R/dataset-places365.R [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    for (n in needed) {
+      res <- self$resources[[n]]
+
+      # Apply extended timeout to all downloads
+      tar <- withr::with_options(list(timeout = 7200), {
+        download_and_cache(res[1], prefix = "places365")
+      })
+
+      if (!is.na(res[2]) && tools::md5sum(tar) != res[2])
+        cli_abort("Corrupt file! Delete the file in {.file {tar}} and try again.")
```
Comment (cregouby) [inline R/dataset-places365.R:199]: **todo** cli_abort message should be _exactly_ the same as existing message or you will have to make a specific translation to each of the variant of the message in each language (see https://github.com/mlverse/torchvision/blob/d6db11996a7092bd453e19844c601338facdf73a/po/R-torchvision.pot#L90-L91) So here please conform to the existing message (i.e. rename `tar` variable into `archive`)

**todo** more generally, for every dataset, you should choose ALL your `cli_` messages within the `po/R-torchvision.pot` file (As per #165)

File: R/dataset-places365.R [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+#'
+#' The dataset is split into:
+#' - `"train"`: images are organized by class folder.
+#' - `"val"`: images are in a flat directory and matched with a label file.
+#' - `"test"`: images are in a flat directory with no labels.
+#'
+#' @inheritParams mnist_dataset
+#' @param split One of `"train"`, `"val"`, or `"test"`.
+#' @param small Logical, defaults to `TRUE`. If `TRUE`, uses the 256 x 256
+#'   #'   "small" images (~30 GB total). If `FALSE`, uses the high-resolution version (~160 GB total).
```
Comment (cregouby) [inline R/dataset-places365.R:13]: **suggestion** Could we make 2 different datasets functions `places365_dataset` and `places365_dataset_large` that would remove the numerous test and boilerplate code for the value of `small`, and would make the code cleaner, and allow a massive inheritance from the first function.

File: R/dataset-places365.R [hunk 11]
```diff
@@ -0,0 +1,225 @@
+#' Places365 Dataset
+#'
+#' Loads the MIT Places365 dataset for scene classification.
+#'
+#' The dataset is split into:
+#' - `"train"`: images are organized by class folder.
+#' - `"val"`: images are in a flat directory and matched with a label file.
+#' - `"test"`: images are in a flat directory with no labels.
```
Comment (cregouby) [inline R/dataset-places365.R:8]: **question** Do the end-user need to know the different folder structure organization ? The goal of the `dataset_` function is explicitely to mask such complexity (but I recognize this is a pain for the developper) ....

File: R/dataset-places365.R (hunk 3 above)
Comment (cregouby) [reply R/dataset-places365.R:161]: **todo** train set must have a `y` value. This connot be avoided.

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** good improvement, and happy to discover that there is a large variant to that dataset
**todo** see inline comments

**Comment by koshtiakanksha** [comment]
I created a new function for the large dataset, but I was wondering if we could move the collate function to a utility module so it can be reused across datasets instead of redefining it each time?

File: R/dataset-places365.R [hunk 12]
```diff
@@ -40,19 +38,32 @@
...
+#'   transform = transform_to_tensor
+#' )
+#' resize_collate <- function(batch) {
+#'   xs <- lapply(batch, function(item) {
+#'     torchvision::transform_resize(item$x, c(256, 256))
+#'   })
+#'   xs <- torch::torch_stack(xs)
+#'   ys <- torch::torch_tensor(sapply(batch, function(it) it$y), dtype = torch::torch_long())
+#'   list(x = xs, y = ys)
+#' }
```
Comment (cregouby) [inline R/dataset-places365.R:54]: **improvement** there is no actual need of the `resize_collate` function is you choose a dataset transform that include a transform resize (see  https://github.com/mlverse/torchvision/blob/8d4fa89b87550df2e414a4e5e99611f6a0a637f0/R/dataset-flowers.R#L25-L28)

File: tests/testthat/test-dataset-places365.R [hunk 13]
```diff
@@ -94,3 +94,113 @@ test_that("places365_dataset test split returns image only", {
...
+  )
+
+  expect_gt(length(ds), 0)
+
+  item <- ds[1]
+  expect_type(item, "list")
+  expect_named(item, c("x", "y"))
+  expect_tensor(item$x)
+  expect_tensor_dtype(item$x, torch::torch_float())
+  expect_true(is.na(item$y))
```
Comment (cregouby) [inline tests/testthat/test-dataset-places365.R:176]: **todo** `y` on a test-set cannot be NA. Please fix the code and update the test.

**Comment by koshtiakanksha** [comment]
Hi @cregouby
 The file, places365_test.txt contains only image filenames and no annotations (labels).
So it does not include class labels, which means item$y will always be NA.

> test_ann_path <- file.path(tempdir(), "places365", "places365_test.txt")
> test_lines <- readLines(test_ann_path, n = 5)
> cat(paste(test_lines, collapse = "\n"))
Places365_test_00000001.jpg
Places365_test_00000002.jpg
Places365_test_00000003.jpg
Places365_test_00000004.jpg
Places365_test_00000005.jpg

What can be done here?

**Comment by cregouby** [comment]
> Hi @cregouby The file, places365_test.txt contains only image filenames and no annotations (labels). So it does not include class labels, which means item$y will always be NA.
> 
> > test_ann_path <- file.path(tempdir(), "places365", "places365_test.txt")
> > test_lines <- readLines(test_ann_path, n = 5)
> > cat(paste(test_lines, collapse = "\n"))
> > Places365_test_00000001.jpg
> > Places365_test_00000002.jpg
> > Places365_test_00000003.jpg
> > Places365_test_00000004.jpg
> > Places365_test_00000005.jpg
> 
> What can be done here?

Oh! Sorry for that. I guess we should mimic the way pytorrch vision did...
